corehq/apps/linked_domain/updates.py

```diff
@@ -543,3 +553,15 @@ def _convert_reports_permissions(domain_link, master_results):
                 new_report_perms.append(get_ucr_class_name(linked_id))
 
         role_def['permissions']['view_report_list'] = new_report_perms
+
+
+def _get_new_tableau_report_permissions(downstream_domain_visualizations, upstream_permissions,
```

Comment:
Given that this is the core of this PR, I'd love to see tests around this method. The tests would also help document the expected behavior, and confirm that the issue with downstream permissions being undesirably overridden would not occur.